packages/lib/gpcircuits/src/multituple.ts

```diff
@@ -0,0 +1,232 @@
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param firstAvailableTupleIndex the index of the first output tuple in the combined array of
+ * entry and tuple value hashes
+ * @param indices the entry value indices forming the input tuple
+ * @returns list of index tuples of arity `paramTupleArity` representing the input tuple
+ * @throws RangeError if the parameters are not in the proper ranges.
+ */
+export function computeTupleIndices(
+  paramTupleArity: number,
+  firstAvailableTupleIndex: number,
+  indices: number[]
+): number[][] {
+  if (indices.some((i) => i >= firstAvailableTupleIndex)) {
+    throw new RangeError(
+      "The first available tuple index must be strictly greater than the given indices."
+    );
+  }
+
+  // Note the first index.
+  const firstIndex = indices[0];
+
+  // The tuple indices will begin with the first `paramTupleArity` indices.
+  const tupleIndices = [indices.slice(0, paramTupleArity)];
+
+  // If we have more than that,
+  if (indices.length > paramTupleArity) {
+    // split the remainder into `paramTupleArity - 1` sized chunks.
+    const unlinkedChunks = toChunks(
+      indices.slice(paramTupleArity),
+      paramTupleArity - 1
+    );
+
+    // Prepend the 'index' of the preceding tuple
+    // to each of these chunks and push them onto `tupleIndices`.
+    unlinkedChunks.forEach((chunk, i) =>
+      tupleIndices.push([firstAvailableTupleIndex + i].concat(chunk))
+    );
+  }
+
+  // Take note of the last tuple for padding purposes.
+  const lastTupleIdx = tupleIndices.length - 1;
+  const lastTuple = tupleIndices[lastTupleIdx];
+  const lastTupleLength = lastTuple.length;
+
+  // Pad the last tuple of `tupleIndices` with `firstIndex` if it is not of length `paramTupleArity` and return.
+  for (let i = 0; i < paramTupleArity - lastTupleLength; i++) {
+    lastTuple.push(firstIndex);
+  }
+
+  return tupleIndices;
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'tuple hash',
+ * thus chaining together smaller tuples to form/represent a larger one.
+ * It accomplishes this by hashing the values, then reducing the resulting
+ * tuple down to one element, `paramTupleArity` elements at a time, via
+ * the Poseidon hash function, padding the tuple with the hash of `elements[0]`
+ * if necessary. The precise mechanism is described in `multiTupleHasher` below.
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param elements the values forming the input tuple
+ * @returns appropriately-formed hash of the input tuple
+ * @throws RangeError if `paramTupleArity` is outside the admissible range
+ */
+export function hashTuple(
+  paramTupleArity: number,
+  elements: PODValue[]
+): bigint {
+  // Call the multituple hasher.
+  const multihash = multiTupleHasher(paramTupleArity, elements);
+
+  // Return the last hash.
+  return multihash[multihash.length - 1];
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'multituple hash',
```

Comment:
Reorder this a bit to separate "what" from "how".  I think the last sentence of this paragraph is the one which belongs first, stating what this function is for.  The rest of the details could be their own paragraph afterwards, or they could be inside the function.  (For this kind of low-level calculation function, I think explaining the algorithm in the doc comment makes sense, but it should come after a description of how to use it.)